beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -100,48 +117,60 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
                 let req = requests
                     .get_mut(&req_id)
                     .ok_or(format!("unknown data columns by range req_id {req_id}"))?;
-                req.finish(req_id, columns)
+                req.finish(req_id, columns, peer_id)
             }
         }
     }
 
-    pub fn responses(&self, spec: &ChainSpec) -> Option<Result<Vec<RpcBlock<E>>, String>> {
-        let Some(blocks) = self.blocks_request.to_finished() else {
+    /// If all internal requests are complete returns a Vec of coupled RpcBlocks
+    #[allow(clippy::type_complexity)]
+    pub fn responses(
+        &self,
+        spec: &ChainSpec,
+    ) -> Option<Result<(Vec<RpcBlock<E>>, BatchPeers), String>> {
+        let Some((blocks, &block_peer)) = self.blocks_request.to_finished() else {
             return None;
         };
 
         match &self.block_data_request {
-            RangeBlockDataRequest::NoData => {
-                Some(Self::responses_with_blobs(blocks.to_vec(), vec![], spec))
-            }
+            RangeBlockDataRequest::NoData => Some(
+                Self::responses_with_blobs(blocks.to_vec(), vec![], spec)
+                    .map(|blocks| (blocks, BatchPeers::new_from_block_peer(block_peer))),
+            ),
             RangeBlockDataRequest::Blobs(request) => {
-                let Some(blobs) = request.to_finished() else {
+                let Some((blobs, _blob_peer)) = request.to_finished() else {
                     return None;
                 };
-                Some(Self::responses_with_blobs(
-                    blocks.to_vec(),
-                    blobs.to_vec(),
-                    spec,
-                ))
+                Some(
+                    Self::responses_with_blobs(blocks.to_vec(), blobs.to_vec(), spec)
+                        .map(|blocks| (blocks, BatchPeers::new_from_block_peer(block_peer))),
+                )
             }
             RangeBlockDataRequest::DataColumns {
                 requests,
-                expected_custody_columns,
+                expected_column_to_peer,
             } => {
                 let mut data_columns = vec![];
+                let mut column_peers = HashMap::new();
```

Comment:
I don't understand why we are rebuilding this hashmap here. Why can't we just pass `expected_column_to_peer` to `BatchPeers::new`

Reply:
`expected_column_to_peer` is the map of columns to peers that we sent request out to. 
`column_peers` is a map of columns we received to the peer ids. 

The difference is `columns_requested_to_peer` vs `columns_received_to_peer`.

Reply:
wouldn't that be the same? the request ids would ensure that right?
not sure why we want to keep 2 separate structures and keep them in sync

Reply:
Yeah true

Reply:
This is what I was thinking https://github.com/pawanjay176/lighthouse/commit/18b2cf3f5aa7ee5dcb1df1bc5a4bd9e126828741

Reply:
This peer tracking mechanism gets replaced anyway in https://github.com/dapplion/lighthouse/pull/45 as we retry individual data_columns_by_range requests. I can use your commit for this intermediary but we'll have to revert it later

Reply:
Yeah i'm fine with us handling that in another PR, just wanted to point out that this PR isn't downscoring the correct peers by itself yet.